### PR TITLE
Implemented the first version of the Spaced Repetition Algorithm.

### PR DESCRIPTION
**Pull request by ImanYZ**

### CODE REVIEW

File: cron/crontab [hunk 1]
```diff
@@ -0,0 +1 @@
+#crontab
```
Comment (presnick) [inline cron/crontab]: What is this file for?

File: inst [hunk 2]
```diff
@@ -0,0 +1,11 @@
+_get_toc_and_questions
```
Comment (presnick) [inline inst]: What is this file for?

File: views/practice/spacedpractice.html [hunk 3]
```diff
@@ -0,0 +1,140 @@
+{{extend 'admin/instructors.html'}}
```
Comment (presnick) [inline views/practice/spacedpractice.html]: I don't see a controller corresponding to this. Did I miss it?

File: cron/crontab (hunk 1 above)
Comment (ImanYZ) [reply cron/crontab]: I have not added this. Maybe it's added by git add.

File: inst (hunk 2 above)
Comment (ImanYZ) [reply inst]: I have not added this. Maybe it's added by git add.

File: views/practice/spacedpractice.html (hunk 3 above)
Comment (ImanYZ) [reply views/practice/spacedpractice.html]: No. I had actually created two controllers for spaced repetition and self-regulated learning. Then you decided to remove the self-regulated and I removed the controller, but forgot to remove this file.

**Comment by presnick** [comment]
Please remove all these unneeded files.

You can do that with `git rm <fname>`

File: controllers/assignments.py [hunk 4]
```diff
@@ -660,8 +695,35 @@ def _autograde_one_q(course_name, sid, question_name, points, question_type, dea
...
         _save_question_grade(sid, course_name, question_name, score, id)
 
+    if practice_start_time:
+        page_visits = db((db.useinfo.course_id == course_name) & \
+                         (db.useinfo.sid == sid) & \
+                         (db.useinfo.event == 'page') & \
+                         (db.useinfo.timestamp >= practice_start_time)) \
+            .select()
+        practice_duration = (datetime.datetime.now() - practice_start_time).seconds / 60
+        practice_score = 0
```
Comment (presnick) [inline controllers/assignments.py]: Please remove the print statements or comment them out.

Comment (ImanYZ) [reply controllers/assignments.py]: In my last commit, I had already removed all the print statements. There are only two of them remaining that are yours.

File: views/practice/index.html [hunk 5]
```diff
@@ -0,0 +1,139 @@
+{{extend 'admin/instructors.html'}}
```
Comment (presnick) [inline views/practice/index.html]: Is this file now moved to views/admin/practice/index.html? If so, please remove this from the git repo with `git rm`

File: views/practice/spacedpractice.html (hunk 3 above)
Comment (presnick) [reply views/practice/spacedpractice.html]: It looks like this file is still in the repo...

File: controllers/assignments.py [hunk 6]
```diff
@@ -590,37 +621,41 @@ def _autograde_one_q(course_name, sid, question_name, points, question_type, dea
             event_filter = 'unittest'
         else:
             event_filter = None
-        results = _scorable_useinfos(course_name, sid, question_name, points, deadline, event_filter)
+        results = _scorable_useinfos(course_name, sid, question_name, points, deadline, event_filter, None,
```
[the remark this answers is not in the document]
Comment (ImanYZ) [reply controllers/assignments.py]: Done.

File: controllers/assignments.py [hunk 7]
```diff
@@ -590,37 +621,41 @@ def _autograde_one_q(course_name, sid, question_name, points, question_type, dea
...
+                                     practice_start_time)
         scoring_fn = _score_one_code_run
     elif question_type == 'mchoice':
-        results = _scorable_mchoice_answers(course_name, sid, question_name, points, deadline)
+        results = _scorable_mchoice_answers(course_name, sid, question_name, points, deadline, practice_start_time)
         scoring_fn = _score_one_mchoice
     elif question_type == 'page':
         # question_name does not help us
-        results = _scorable_useinfos(course_name, sid, question_name, points, deadline, question_type='page')
+        results = _scorable_useinfos(course_name, sid, question_name, points, deadline, None, 'page',
```
[the remark this answers is not in the document]
Comment (ImanYZ) [reply controllers/assignments.py]: Done.

**Comment by ImanYZ** [comment]
I removed all the unnecessary files.

--
Iman YeckehZaare
Ph.D. Candidate
University of Michigan School of Information

[image: http://www.iman.one/] <http://www.iman.one/>   [image:
https://www.linkedin.com/in/oneweb/] <https://www.linkedin.com/in/oneweb/>
 [image: https://stackoverflow.com/users/2521204/iman]
<https://stackoverflow.com/users/2521204/iman>   [image:
https://en.wikipedia.org/wiki/User:I.yeckehzaare]
<https://en.wikipedia.org/wiki/User:I.yeckehzaare>   [image:
https://www.si.umich.edu/people/iman-yeckehzaare]
<https://www.si.umich.edu/people/iman-yeckehzaare>

**Comment by presnick** [comment]
This has been merged to branch feature_practice.

**Pull request merged**